crates/apollo-compiler/src/name.rs

```diff
@@ -0,0 +1,511 @@
+        let len = value.len();
+        if len >= (u32::MAX as usize) {
+            panic!("Name length overflows 4 GiB")
+        }
+        len as _
+    }
+
+    pub fn location(&self) -> Option<NodeLocation> {
+        let file_id = self.tagged_file_id.file_id();
+        if file_id != FileId::NONE {
+            Some(NodeLocation {
+                file_id,
+                text_range: TextRange::at(self.start_offset.into(), self.len.into()),
+            })
+        } else {
+            None
+        }
+    }
+
+    /// If this string contains a location, convert it to line and column numbers
+    pub fn line_column(&self, sources: &SourceMap) -> Option<GraphQLLocation> {
+        GraphQLLocation::from_node(sources, self.location())
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.len as _
+    }
+
+    #[inline]
+    pub fn as_str(&self) -> &str {
+        let slice = NonNull::slice_from_raw_parts(self.ptr, self.len());
+        // SAFETY: all constructors set `self.ptr` and `self.len` from valid UTF-8,
+        // and we return a lifetime tied to `self`.
+        unsafe { std::str::from_utf8_unchecked(slice.as_ref()) }
+    }
+
+    /// If this `Name` was created with [`new_static`][Self::new_static]
+    /// or the [`name!`][crate::name] macro, return the string with `'static` lifetime.
+    pub fn as_static_str(&self) -> Option<&'static str> {
+        if self.tagged_file_id.tag() == TAG_STATIC {
+            let raw_slice = NonNull::slice_from_raw_parts(self.ptr, self.len());
+            // SAFETY: the tag indicates `self.ptr` came from `Self::ptr_and_tag_from_static`,
+            // so it has the static lifetime and points to valid UTF-8 of the correct length.
+            Some(unsafe { std::str::from_utf8_unchecked(raw_slice.as_ref()) })
+        } else {
+            None
+        }
+    }
+
+    fn as_arc(&self) -> Option<ManuallyDrop<Arc<str>>> {
+        if self.tagged_file_id.tag() == TAG_ARC {
+            let raw_slice = NonNull::slice_from_raw_parts(self.ptr, self.len())
+                .as_ptr()
+                .cast_const();
+
+            // SAFETY:
+            //
+            // * The tag indicates `self.ptr` came from `Arc::into_raw` in `ptr_and_tag_with_arc`
+            // * `Arc::from_raw` normally moves ownership away from the raw pointer,
+            //   `ManuallyDrop` counteracts that
+            Some(ManuallyDrop::new(unsafe {
+                Arc::from_raw(raw_slice as *const str)
+            }))
+        } else {
+            None
+        }
+    }
+
+    /// If this `Name` was created with [`new_static`][Self::new_static]
+    /// or the [`name!`][crate::name] macro, return the string with `'static` lifetime.
+    ///
+    /// Otherwise, return a clone of the `Arc` used internally for this `Name`.
+    pub fn to_static_str_or_cloned_arc(&self) -> Result<&'static str, Arc<str>> {
```

Comment:
yeaaaaaa i think this needs to be a enum for a returned type. A bit annoying with the various `str`s, I know. Just doesn't feel right to use `Result` in a public API for something that isn't one.